### PR TITLE
rephrase UnsafeCell doc

### PR DESCRIPTION
**Pull request by NovemberZulu**
As shown by discussions on users.rust-lang.org [[1]], [[2]], UnsafeCell doc is not totally clear. I tried to made the doc univocal regarding what is allowed and what is not. The edits are based on my understanding following [[1]].

[1]: https://users.rust-lang.org/t/unsafecell-behavior-details/1560
[2]: https://users.rust-lang.org/t/is-there-a-better-way-to-overload-index-indexmut-for-a-rc-refcell/15591/12

### CODE REVIEW

**Comment by TimNN** [comment]
Re-assigning to docs team.

r? @steveklabnik

File: src/libcore/cell.rs [hunk 1]
```diff
@@ -1164,11 +1164,12 @@ impl<'a, T: ?Sized + fmt::Display> fmt::Display for RefMut<'a, T> {
 /// The compiler makes optimizations based on the knowledge that `&T` is not mutably aliased or
 /// mutated, and that `&mut T` is unique. When building abstractions like `Cell`, `RefCell`,
 /// `Mutex`, etc, you need to turn these optimizations off. `UnsafeCell` is the only legal way
-/// to do this. When `UnsafeCell<T>` is immutably aliased, it is still safe to obtain a mutable
-/// reference to its interior and/or to mutate it. However, it is up to the abstraction designer
-/// to ensure that no two mutable references obtained this way are active at the same time, and
-/// that there are no active mutable references or mutations when an immutable reference is obtained
-/// from the cell. This is often done via runtime checks.
+/// to do this. When `UnsafeCell<T>` _itself_ is immutably aliased, it is still safe to obtain
+/// a mutable reference to its _interior_ and/or to mutate the interior. However, it is up to
```
Comment (GuillaumeGomez) [inline src/libcore/cell.rs]: I'm not sure that putting italic in here brings anything.

**Comment by GuillaumeGomez** [comment]
Except for my little concern, it seems good to me.

**Comment by steveklabnik** [comment]
I would really like @rust-lang/libs or @rust-lang/lang to sign off here; changing an unsafe lang item's docs is something i'm super wary of doing without their input.

File: src/libcore/cell.rs (hunk 1 above)
Comment (NovemberZulu) [reply src/libcore/cell.rs]: removed italic

File: src/libcore/cell.rs [hunk 2]
```diff
@@ -1164,11 +1164,12 @@ impl<'a, T: ?Sized + fmt::Display> fmt::Display for RefMut<'a, T> {
 /// The compiler makes optimizations based on the knowledge that `&T` is not mutably aliased or
 /// mutated, and that `&mut T` is unique. When building abstractions like `Cell`, `RefCell`,
 /// `Mutex`, etc, you need to turn these optimizations off. `UnsafeCell` is the only legal way
-/// to do this. When `UnsafeCell<T>` is immutably aliased, it is still safe to obtain a mutable
-/// reference to its interior and/or to mutate it. However, it is up to the abstraction designer
-/// to ensure that no two mutable references obtained this way are active at the same time, and
-/// that there are no active mutable references or mutations when an immutable reference is obtained
-/// from the cell. This is often done via runtime checks.
+/// to do this. When `UnsafeCell<T>` itself is immutably aliased, it is still safe to obtain
+/// a mutable reference to its interior and/or to mutate the interior. However, it is up to
```
Comment (cramertj) [inline src/libcore/cell.rs]: Can you reformat this as a list? Several of the entries are long, and the sentence becomes hard to follow.

File: src/libcore/cell.rs [hunk 3]
```diff
@@ -1164,11 +1164,12 @@ impl<'a, T: ?Sized + fmt::Display> fmt::Display for RefMut<'a, T> {
...
+/// the abstraction designer to ensure that no two mutable references obtained this way are active
+/// at the same time, there are no active immutable reference when a mutable reference is obtained
```
Comment (cramertj) [inline src/libcore/cell.rs]: These three rules are really all just one rule: while a mutable reference exists, no second reference of any kind may exist.

Comment (NovemberZulu) [reply src/libcore/cell.rs]: I'll try to merge three rules into one, so it will be easier to follow.

**Comment by NovemberZulu** [comment]
As far as I understand -- and I quite possible I am wrong -- simply storing a "bad" reference is not a crime, it is just some bytes in memory, but actually using it is UB, so I talk about active references in the doc. What exactly is "using" as opposed to stored is another story, and should be defined in memory model, I believe.

**Comment by nikomatsakis** [comment]
I tried to offer some feedback but found it harder than expected. I will say that I continue to be strongly opposed to the term "immutable reference" -- I know we use it a lot, but **particularly in the context of UnsafeCell**, it strikes me as quite confusing. I'd prefer that we either say "shared reference" or just `&`-reference (perhaps clearer). 

I personally found the current form of "English-y paragraph" sort of confusing, to be honest. I think I'd prefer something that says like:

`UnsafeCell` is the only way that data reached via a `&`-reference can be correctly mutated: any data that is not stored within an `UnsafeCell` must be immutable when reachable via `&` reference. The `UnsafeCell` API itself is very simple: it gives you a raw pointer (`*mut T`) to its contents. It is up to *you* to use that raw pointer correctly.

The precise Rust aliasing rules are somewhat in flux, but the main points are not contentious:

- If you create a safe reference with lifetime `'a` (either a `&` or `&mut` reference) that is accessible by safe code (for example, because you returned it), then you must not access the data in any way that contradicts that reference for the remainder of `'a`.
  - For example, that means that if you take the `*mut T` from an `UnsafeCell<T>` and case it to an `&T`, then until that reference's lifetime expires, the data in `T` must remain immutable (modulo any `UnsafeCell` data found within `T`, of course). Similarly, if you create an `&mut T` reference that is released to safe code, then you must not access the data within the `UnsafeCell` until that reference expires.
- At all times, you must avoid data races, meaning that if multiple threads have access to the same `UnsafeCell`, then any writes must have a proper happens-before relation to all other accesses (or use C++11 atomics).

**Comment by NovemberZulu** [comment]
I totally agree that proper wording here is a challenge -- the text must be univocal, concise, yet as simple as possible. I think that "shared reference" might be a bit confusing, because of "shared with what?" question. `&`-reference is a bit jargon term, but `&`-reference contrasts `&mut`-reference very nicely.

I think the best way is to state explicitly allowed code flows, at least for single-thread access -- multi-thread access is way more complicated (does Rust strictly define  happens-before relation?) -- something like this:

`UnsafeCell` is the only way in core Rust that data reached via a `&`-reference can be correctly mutated: any data that is not stored within an `UnsafeCell` must be immutable when reachable via `&`-reference. All other types that allow internal mutability, such as `Cell<T>` and `RefCell<T>` use `UnsafeCell` to wrap their internal data.  The `UnsafeCell` API itself is very simple: it gives you a raw pointer (`*mut T`) to its contents. It is up to _you_ as the abstraction designer to use that raw pointer correctly.

The precise Rust aliasing rules are somewhat in flux, but the main points are not contentious:

- If you create a safe reference with lifetime `'a` (either a `&` or `&mut` reference) that is accessible by safe code (for example, because you returned it), then you must not access the data in any way that contradicts that reference for the remainder of `'a`. For example, that means that if you take the `*mut T` from an `UnsafeCell<T>` and case it to an `&T`, then until that reference's lifetime expires, the data in `T` must remain immutable (modulo any `UnsafeCell` data found within `T`, of course). Similarly, if you create an `&mut T` reference that is released to safe code, then you must not access the data within the `UnsafeCell` until that reference expires. The following two scenarios are explicitly legal:

    1. a `&`-reference can be released to safe code and there it can co-exit with other `&`-references, but not with a `&mut`-reference

    2. a `&mut`-reference may be released to safe code, provided neither other `&mut`-references nor even `&`-references co-exist with it. A `&mut`-reference must always be unique.

- At all times, you must avoid data races, meaning that if multiple threads have access to the same `UnsafeCell`, then any writes must have a proper happens-before relation to all other accesses (or use atomics).

**Comment by nikomatsakis** [comment]
> does Rust strictly define happens-before relation

We inherit the C++11 model, basically.

**Comment by NovemberZulu** [comment]
> We inherit the C++11 model, basically.

(sigh) I was so hoping to get away from it...

**Comment by cmr** [comment]
@NovemberZulu why? because of how complex it is?

**Comment by NovemberZulu** [comment]
Because, as with numerous other C++ features, you never know when it is going to bite you in the ass

**Comment by cmr** [comment]
I'm confused. The C++11 memory model is extremely well understood. http://www.cl.cam.ac.uk/~pes20/weakmemory/#CPP

**Comment by NovemberZulu** [comment]
C++ is hardly an esoteric or obscure  language, and the documentation is both readily available and extensive, that is certainly true. Still, to my mind, it is too easy to end up with a code that looks OK, seems to work the way you expect it to work, but is, in fact, UB.

P.S. I think we are digressing.

**Comment by steveklabnik** [comment]
@nikomatsakis we use "reference" and "mutable reference", officially. Or at least, that's what I've been keeping consistent in every bit of docs I come across, it's hard, of course.

I agree "immutable reference" is a bit misleading, which is part of why we chose this this way so long ago.

**Comment by NovemberZulu** [comment]
> we use "reference" and "mutable reference", officially. 

I am afraid I have been using "reference"  as an umbrella term for both `&T` and `&mut T` adding to the confusion :( Following Liskov, `&mut T` IS-A `&T`, isn't? BTW, what is misleading about "immutable reference"?

Anyway, may be we can just use `&T` and `&mut T` in `UnsafeCell<T>` doc?

**Comment by Kimundi** [comment]
@NovemberZulu: A `&T` might allow some mutation inside the `T` if `T` uses internal mutability like `Cell`, so its not always immutable.

**Comment by NovemberZulu** [comment]
> A `&T` might allow some mutation inside the `T` if `T` uses internal mutability like `Cell`, so its not always immutable.

Yes, true, and this is exactly the case we have.  I still have a mental picture of `T * const` when thinking of `UnsafeCell<T>` :(

**Comment by nikomatsakis** [comment]
@steveklabnik 

> we use "reference" and "mutable reference", officially. Or at least, that's what I've been keeping consistent in every bit of docs I come across, it's hard, of course.

That make sense, but it seems like sometimes we will want to clarify that we mean an `&T` and that we are excluding `&mut T`. In those cases, what do you do?

**Comment by steveklabnik** [comment]
If "reference" isn't clear enough, I often write `&T`.

**Comment by NovemberZulu** [comment]
Another attempt at rewording:

If you have a reference `&SomeStruct`, then normally in Rust all fields of `SomeStruct` are immutable.  `UnsafeCel<T>` is the only core language feature to work around this restriction.  All other types that allow internal mutability, such as `Cell<T>` and `RefCell<T>` use `UnsafeCell` to wrap their internal data.  The `UnsafeCell` API itself is technically very simple: it gives you a raw pointer `*mut T` to its contents. It is up to _you_ as the abstraction designer to use that raw pointer correctly.

The precise Rust aliasing rules are somewhat in flux, but the main points are not contentious:

- If you create a safe reference with lifetime `'a` (either a `&T` or `&mut T` reference) that is accessible by safe code (for example, because you returned it), then you must not access the data in any way that contradicts that reference for the remainder of `'a`. For example, that means that if you take the `*mut T` from an `UnsafeCell<T>` and case it to an `&T`, then until that reference's lifetime expires, the data in `T` must remain immutable (modulo any `UnsafeCell` data found within `T`, of course). Similarly, if you create an `&mut T` reference that is released to safe code, then you must not access the data within the `UnsafeCell` until that reference expires. 

- At all times, you must avoid data races, meaning that if multiple threads have access to the same `UnsafeCell`, then any writes must have a proper happens-before relation to all other accesses (or use atomics).

To assist with proper design, the following scenarios are explicitly declared legal for single-thread code:

1. A `&T` reference can be released to safe code and there it can co-exit with other `&T` references, but not with a `&mut T` 

2. A `&mut T` reference may be released to safe code, provided neither other `&mut T` nor `&T` co-exist with it. A `&mut T` must always be unique.

**Comment by nikomatsakis** [comment]
@NovemberZulu I like it quite a lot

**Comment by NovemberZulu** [comment]
Updated the pull request.  I feel the transition between old and new text is bit rough. Comments are welcome!

**Comment by pietroalbini** [comment]
Ping from triage @steveklabnik! The author pushed new commits, could you review them?

**Review by steveklabnik — APPROVED** [review]

**Comment by steveklabnik** [comment]
I think it reads great, thanks!

@bors: r+ rollup

**Comment by NovemberZulu** [comment]
Everyone, thank you very much for your support!
